### PR TITLE
feat: add autoupdate-workflow

### LINKED ISSUE #227: Try to improve autoupdate
#### Reason/Context

The [autoupdate workflow](https://github.com/asyncapi/.github/blob/master/.github/workflows/autoupdate.yml) doesn't work with forks. 

It can update only branches that are in the upstream.

#### Description

When the branch from the fork is out of sync with master, in the UI maintainers see a button that they can click to update the branch with master.

We should allow anyone to be able to trigger such an update.

If there is a button, there needs to be API behind it, REST or GraphQL. There needs to be a research on how to trigger such branch update in form with an API. Then we need to update autoupdate workflow

### PR DESCRIPTION
**Pull request by Shurtu-gal**
**Description**

- Replaces previous auto-update workflow with the current one.
- Uses an action developed by me :- https://github.com/marketplace/actions/autoupdate-fork-action
- The action currently runs on push to branches or `/update` comment and updates PRs pointing to it and having `autoupdate` label in it.
- Update `/help` to have `/update` in it.

**Related issue(s)**
Fixes #227

### CODE REVIEW

**Review by derberg — CHANGES_REQUESTED** [review]
I'm not 100% sure how it works, also please share examples where you tested it.

I'm confused because:
- I do not understand why autoupdate workflow had to be removed
- what is the difference between `update` and `autoupdate` (cause you left autoupdate). Is it that `autoupdate` you use once and then it is always triggered, and `update` you need to pass every time you wanna trigger update?
- in short explain why previous action did not work and what technical change you do in your action that now bot token will be able to push to fork

**Comment by Shurtu-gal** [comment]
@derberg as was discussed over slack, to not consume too much action minutes right now auto-update is only enabled for comments like `\u` or `/update`. Hence there is no need of filtering based on labels.

[Successful run](https://github.com/Shurtu-gal/autoupdate-fork-action/actions/runs/7992039025) [The related PR](https://github.com/Shurtu-gal/autoupdate-fork-action/pull/1)

Labels were meant for filtering in case of mass updates, [like in this case](https://github.com/Shurtu-gal/autoupdate-fork-action)

Also, I had removed the previous one as I this one was sufficient for both tasks, but before we do so we have to consider amount of action minutes it will take. Furthermore, GH API limit may also be reached.

### Why wasn't the previous one working?
Because the API used by it requires admin access for the updation. Meanwhile, I am [this mutation](https://docs.github.com/en/graphql/reference/mutations#updatepullrequestbranch) which basically uses a merge PR already created by GitHub (Yeah, I know it sounds unreal 😆), and applies it on top of the HEAD, this was custom made for PR branches.

**Comment by derberg** [comment]
the problem with saving minutes is that we have a huge amount of bot-created PRs, that quite often require autoupdating

one example: https://github.com/asyncapi/asyncapi-react/pull/924 -> by default autoupdate is added and the bot merges master in like a charm

**Comment by Shurtu-gal** [comment]
@derberg 
A successful run on comment:- https://github.com/ash17290/asyncapi-github/actions/runs/8086468906
A Successful run on push:- https://github.com/ash17290/asyncapi-github/actions/runs/8086440737

Note that only labelled PRs are auto-updated rest have to be manually done.

File: .github/workflows/automerge-for-humans-add-ready-to-merge-or-do-not-merge-label.yml [hunk 1]
```diff
@@ -59,7 +59,7 @@ jobs:
                 body: `Hello, @${{ github.actor }}! 👋🏼
                        This PR is not up to date with the base branch and can't be merged.
                        Please update your branch manually with the latest version of the base branch.
-                       PRO-TIP: Add a comment to your PR with the text: \`/au\` or \`/autoupdate\` and our bot will take care of updating the branch in the future. The only requirement for this to work is to enable [Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) option in your PR.
+                       PRO-TIP: Add a comment to your PR with the text: \`/u\` or \`/update\` and our bot will take care of updating the branch in the future. The only requirement for this to work is to enable [Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) option in your PR.
```
Comment (derberg) [inline .github/workflows/automerge-for-humans-add-ready-to-merge-or-do-not-merge-label.yml:62]:
```suggestion
                       PRO-TIP: To request an update from the upstream branch, simply comment \`/u\` or \`/update\` and our bot will handle the update operation promptly.
                       
                       The only requirement for this to work is to enable [Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) option in your PR. Also the update will not work if your fork is located in an organization, not under your personal profile.
```

**Comment by Shurtu-gal** [comment]
@derberg The refactoring has been done.

File: .github/workflows/autoupdate-fork.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+on:
+  issue_comment: 
+    types: [created]
+
+jobs:
+  autoupdate-pr:
+    if: >
+      startsWith(github.repository, 'asyncapi/') &&
+      github.event_name == 'issue_comment' && (
```
Comment (derberg) [inline .github/workflows/autoupdate-fork.yml:20]: this is always this even name, 

instead you need something like:
```
github.event.issue.pull_request && 
      github.event.issue.state != 'closed' && 
```

example: https://github.com/asyncapi/.github/blob/master/.github/workflows/automerge-for-humans-add-ready-to-merge-or-do-not-merge-label.yml#L18-L19

File: .github/workflows/autoupdate.yml [hunk 3]
```diff
@@ -1,34 +1,34 @@
...
-    steps:
-      - name: Autoupdating
-        uses: docker://chinthakagodawita/autoupdate-action:v1
-        env:
-          GITHUB_TOKEN: '${{ secrets.GH_TOKEN_BOT_EVE }}'
-          PR_FILTER: "labelled"
-          PR_LABELS: "autoupdate"
-          PR_READY_STATE: "ready_for_review"
-          MERGE_CONFLICT_ACTION: "ignore"
+# This action is centrally managed in https://github.com/asyncapi/.github/
```
[the remark this answers is not in the document]
Comment (Shurtu-gal) [reply .github/workflows/autoupdate.yml:1]: Yeah, the git hashes may be different, there has been no changes though.

File: .github/workflows/autoupdate-fork.yml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+              core.debug(GraphQLError);
+              if (
+                GraphQLError.name === 'GraphqlResponseError' &&
+                GraphQLError.errors.some(
+                  error => error.type === 'FORBIDDEN' || error.type === 'UNAUTHORIZED'
+                )
+              ) {
+                // Add comment to PR if the bot doesn't have permissions to update the PR
+
+                const comment = `@${context.actor} I don't have permissions to update this PR. Please ask the PR author to check the "Allow edits from maintainers" checkbox in the PR settings.`;
```
[the remark this answers is not in the document]
Comment (Shurtu-gal) [reply .github/workflows/autoupdate-fork.yml]: Added it.

File: .github/workflows/help-command.yml [hunk 5]
```diff
@@ -33,7 +33,8 @@ jobs:
         
                     - \`/ready-to-merge\` or \`/rtm\` - This comment will trigger automerge of PR in case all required checks are green, approvals in place and do-not-merge label is not added
                     - \`/do-not-merge\` or \`/dnm\` - This comment will block automerging even if all conditions are met and ready-to-merge label is added
-                    - \`/autoupdate\` or \`/au\` - This comment will add \`autoupdate\` label to the PR and keeps your PR up-to-date to the target branch's future changes. Unless there is a merge conflict or it is a draft PR.`
+                    - \`/autoupdate\` or \`/au\` - This comment will add \`autoupdate\` label to the PR and keeps your PR up-to-date to the target branch's future changes. Unless there is a merge conflict or it is a draft PR. (Currently only works for upstream branches.)`
+                    - \`/update\` or \`/u\` - This comment will update the PR with the latest changes from the target branch. Unless there is a merge conflict or it is a draft PR.`
```
[the remark this answers is not in the document]
Comment (Shurtu-gal) [reply .github/workflows/help-command.yml]: Added.

File: .github/workflows/autoupdate-fork.yml [hunk 6]
```diff
@@ -0,0 +1,99 @@
+# This workflow is centrally managed in https://github.com/asyncapi/.github/
+# Don't make changes to this file in this repo as they will be overwritten with changes made to the same file in above mentioned repo
+
+# This workflow will run on every comment with /update or /u. And will create merge-commits for the PR.
+# This also works with forks, not only with branches in the same repository/organization.
+# Currently, does not work with forks in different organizations.
+
+# This workflow will be distributed to all repositories in the AsyncAPI organization
+
+name: Autoupdate PR branches from fork
```
Comment (derberg) [inline .github/workflows/autoupdate-fork.yml:10]: don't want to be picky but we need to handle naming here with caution so we do not confuse it with autoupdate command

this one should be `Update PR branches from fork`, same filename and job name like `autoupdate-pr` should be `update-pr`

File: .github/workflows/autoupdate-fork.yml [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+              core.debug(GraphQLError);
+              if (
+                GraphQLError.name === 'GraphqlResponseError' &&
+                GraphQLError.errors.some(
+                  error => error.type === 'FORBIDDEN' || error.type === 'UNAUTHORIZED'
+                )
+              ) {
+                // Add comment to PR if the bot doesn't have permissions to update the PR
+
+                const comment = `@${context.actor} I don't have permissions to update this PR. Please ask the PR author to check the "Allow edits from maintainers" checkbox in the PR settings. Another possible reason for failure can due to the fork being in an organisation and not under a user.`;
```
[the remark this answers is not in the document]
Comment (Shurtu-gal) [reply .github/workflows/autoupdate-fork.yml:87]: Thanks will do so.

**Comment by Shurtu-gal** [comment]
> also in error messages for user, we do not write anything about merge conflicts, that in case of merge conflicts it will also not work

Added that as well.

**Review by derberg — APPROVED** [review]
great stuff!!!

**Pull request merged**